hashicorp/dev-portal
# unified docs: add per-product docs loader flag

## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

This PR introduces a feature flag for migrating to Unified Docs.

## 🤷 Why

To allow incremental migration to our Unified Docs Repository, specifically by allowing content source repositories to be migrated individually. This continues the work started in #2550, adding feature flags to allow use of the `UNIFIED_DOCS_API` `.env` variable introduced in that PR.

## 🛠️ How

- Introduces a `unified_docs_migrated_repos` flag, which is expected to be an array of strings representing content source repositories
    - Note that "content source repository" is distinct from "product". For example, the `terraform` product has content sourced from multiple repositories, such as `terraform-docs-common`, `ptfe-releases`, `terraform-plugin-mux`, and so on.
- Builds a utility function `getContentApiBaseUrl(repoName)`, which makes a relatively simple check against the `unified_docs_migrated_repos` flag
    - Splitting out this utility felt intuitive, as in the first few use cases for the flag, it became clear there'd be identical logic since the situation is usually "we're about to make a content API call in the context of some repo, which URL should we make the request to?"
    - Splitting this out also provided a space to add some detailed comments on caveats and limitations related to the unified docs migration so far. In brief, this flag currently represents the migration of "fetching documents", but there are other yet-to-be-implemented content API functions that may make sense to migrate at the same time as "fetching documents". I've put as much detail as I thought might be helpful in comments within the utility file that expand on this.
- Used `getContentApiBaseUrl` utility function in docs-view remote loader
    - If a content source repo has been migrated, we should fetch `version-metadata`, `nav-data`, and `doc` documents from the unified docs API.
- Used `getContentApiBaseUrl` utility function in get-valid-versions
    - If a content source repo has been migrated, we should fetch version metadata on each document from the unified docs API via `/api/content-versions`. Note that this route has [yet to be fully implemented](https://app.asana.com/0/1207899860738460/1208471953509347/f).

Related but less consequential changes:

- Updates `.env` UNIFIED_DOCS_API to remove trailing slash, for consistency

## 📸 Design Screenshots

![CleanShot 2024-10-18 at 11 37 23@2x](https://github.com/user-attachments/assets/22d8a524-d713-4e97-a558-370fadb4c8d5)

## 🧪 Testing

- Check out [hashicorp/web-unified-docs](https://github.com/hashicorp/web-unified-docs), and run `npm i && npm run prebuild && npm run dev` to start the API locally at `http://localhost:8080`
- Check out [hashicorp/dev-portal](https://github.com/hashicorp/dev-portal) to this branch, `zs.add-unified-docs-document-flag`. Modify the `.env` file so that `UNIFIED_DOCS_API= http://localhost:8080`, then modify `config/development.json` to add a flagged content source repo, for example `flags: { "unified_docs_migrated_repos": ["terraform"] }`. Then run `npm i && npm run start`, and visit the docs associated with the flagged content source repo. These docs should be served from the locally running `web-unified-docs` instance. All other docs should be served from our existing content API (`content.hashicorp.com`).

## 💭 Anything else?

Not at the moment!

[task]: https://app.asana.com/0/799513369421962/1208471953509340/f
[preview]: https://dev-portal-git-zsadd-unified-docs-document-flag-hashicorp.vercel.app/
